services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,428 @@
+                        })
+                        .orElseGet(() -> MongoDbDeviceRegistryUtils.checkForVersionMismatchAndFail(tenantId,
+                                resourceVersion, findTenant(tenantId))));
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final String tenantId) {
+        Objects.requireNonNull(tenantId);
+
+        return get(tenantId, NoopSpan.INSTANCE);
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(span);
+
+        return readTenant(tenantId, span)
+                .compose(tenantDtoResult -> {
+                    if (tenantDtoResult.getStatus() != HttpURLConnection.HTTP_OK) {
+                        TracingHelper.logError(span, "tenant not found");
+                        return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_NOT_FOUND));
+                    }
+                    return Future.succeededFuture(TenantResult.from(
+                            HttpURLConnection.HTTP_OK,
+                            DeviceRegistryUtils
+                                    .convertTenant(tenantId, tenantDtoResult.getPayload(), true),
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())));
+                });
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn) {
+        Objects.requireNonNull(subjectDn);
+
+        return get(subjectDn, NoopSpan.INSTANCE);
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn, final Span span) {
+        Objects.requireNonNull(subjectDn);
+        Objects.requireNonNull(span);
+
+        return getForCertificateAuthority(subjectDn, span);
+    }
+
+    /**
+     * Finds an existing tenant by the subject DN of its configured certificate authority.
+     *
+     * @param subjectDn the subject DN to find the tenant with.
+     * @param span The active OpenTracing span for this operation. It is not to be closed in this method!
+     *            An implementation should log (error) events on this span and it may set tags
+     *            and use this span as the parent for any spans created in this method.
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>200 OK</em> if a tenant with the given certificate authority is registered.
+     *         The <em>payload</em> will contain the tenant's configuration information.</li>
+     *         <li><em>400 Bad Request</em> if the certificate authority is missing.</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier.</li>
+     *         </ul>
+     * @throws NullPointerException if the parameters are {@code null}.
```

Comment:
I don't think this is correct, as in case of a `null` value in the `subjectDN` parameter, the result would not be a NPE, but a proper result with `HTTP_BAD_REQUEST`.

Reply:
I have oriented myself to another file base service, but throwing NPE makes little sense when it is checked afterwards.
In the inheriting method [get](https://github.com/bosch-io/hono/blob/b385b0ad53efdb0cbcf66fd04038c02dadb05989/service-base/src/main/java/org/eclipse/hono/service/tenant/TenantService.java#L91), NPE is expected to be present when `subjectDn == null`. Should `HTTP_BAD_REQUEST` be returned at this point?

Reply:
I think it should either be a NPE or `HTTP_BAD_REQUEST` but not both. And the javadoc should document what is happening.

My personal preference would be to throw a `ServiceInvocationException` with the status code of `HTTP_BAD_REQUEST`. This would give you information what happened, and a stack trace of where.

Reply:
I suggest to remove the NPE checks and return `HTTP_BAD_REQUEST` for all methods that inherit from `TenantService`.

Reply:
Further discussed in https://github.com/eclipse/hono/pull/1845#discussion_r421352466